File: dataAnalysis/matlab/ReadInYahooFinanceData.m
function ReadInYahooFinanceData(dir,filename,SaveFile, ...
                                year0,month0,day0, ...
                                year1,month1,day1)
  fprintf('> Loading in symbols from %s \n',[dir,filename]);
  
  fid = fopen([dir,filename]);
  k = 1;
  while 1
    tline = fgetl(fid);
    if ~ischar(tline)
      break
    else
      if tline(1) ~= '#'
        symbols{k} = tline;
        k = k+1;
      end
    end
  end
  fclose(fid);

  Ns = length(symbols);
  data = cell(1,Ns);
  
  fprintf('> Read %d symbols \n',Ns);

  % t1 = datetime(year0,month0,day0);
  % t2 = datetime(year1,month1,day1);
  
  % day0_key = posixtime(t1);
  % day1_key = posixtime(t2);
  
  %% get data from yahoo finance
  %  (1),  (2),  (3), (4),   (5),    (6),       (7)
  % date, open, high, low, close, volume, adj close
  for k = 1:Ns

    % url = sprintf('%s%s%s%d%s%d%s%d%s%d%s%d%s%d%s', ...
    %               'https://chart.finance.yahoo.com/table.csv?s=', ...
    %               symbols{k}, ...
    %               '&a=',month0,'&b=',day0,'&c=',year0, ...
    %               '&d=',month1,'&e=',day1,'&f=',year1,'&g=d&ignore=.csv');
    % url = sprintf(['%s%s?period1=%d&period2=%d&interval=1d&events=' ...
    %                'history&crumb=E.z47grLd4d'], ...
    %               'https://query1.finance.yahoo.com/v7/finance/download/', ...
    %               symbols{k},day0_key,day1_key);

    file_name = '/home/dan/tabletest.csv';
    [status,result] = ...
        unix(['python ', ...
              '/home/dan/tradeStratsAndStats/dataAnalysis/matlab/ReadInYahooFinanceData.py ', ...
              sprintf('%s %d %d %d %d %d %d', ...
                      symbols{k},year0,month0,day0, ...
                      year1,month1,day1)]);
    if ~isempty(result)
      symbols{k}
      error(result);
    end

    data{k} = csvread(file_name,1,1);
    tmp_ = data{k}(:,5);
    data{k}(:,5) = data{k}(:,6);
    data{k}(:,6) = tmp_;

    N = length(data{k}(:,1));

    
    fid = fopen(file_name);
    l = N;
    tline = fgetl(fid);
    while 1
      tline = fgetl(fid);
      if ~ischar(tline)
        break
      else
        tmp = textscan(tline,'%10s\n');
        dates{k}{l} = tmp{1};
        l = l-1;
      end
    end
    fprintf('> loaded %s [%d/%d] \n',symbols{k},k,Ns);
  end
  
  N = length(data{1}(:,1));
  save([dir,SaveFile],'data','dates','Ns','N','symbols');
  fprintf('> Saved data to %s \n',[dir,SaveFile]);
end

File: dataAnalysis/matlab/ReadInYahooFinanceData.py
#!/usr/bin/python
import sys
import base64
import urllib
import urllib2
import time
import datetime
import ssl

if __name__ == '__main__':
    symbol = sys.argv[1]
    year0  = eval(sys.argv[2])
    month0 = eval(sys.argv[3])
    day0   = eval(sys.argv[4])
    year1  = eval(sys.argv[5])
    month1 = eval(sys.argv[6])
    day1   = eval(sys.argv[7])


    # convert to date time
    t1 = datetime.datetime(year0,month0,day0,0,0)

    # convert to date time
    t2 = datetime.datetime(year1,month1,day1,0,0)

    # get time in seconds based on standard of 1/1/70
    day0_key = (t1-datetime.datetime(1970,1,1)).total_seconds()
    day1_key = (t2-datetime.datetime(1970,1,1)).total_seconds()

    url = "%s%s?period1=%d&period2=%d&interval=1d&events=history&crumb=E.z47grLd4d" % ( \
              'https://query1.finance.yahoo.com/v7/finance/download/', \
              symbol, \
              day0_key,day1_key)

    # print url
    ### Deal with api access key
    client_id = "dj0yJmk9dXVURnhKZVJtbXRVJmQ9WVdrOVExYzNkbTVKTldNbWNHbzlNQS0tJnM9Y29uc3VtZXJzZWNyZXQmeD1iYQ--"
    client_secret = "98767d5780570c652504bee46626de23b1d294a1"
    credentials = "%s:%s" % (client_id, client_secret)
    encode_credential = base64.b64encode(credentials.encode('utf-8')).decode('utf-8').replace("\n", "")
    header_params = {
        "Authorization": ("Basic %s" % encode_credential),
        "Content-Type": "application/x-www-form-urlencoded",
        "Accept": "application/json"
    }
    param = {
        'grant_type': 'client_credentials',
    }
    data = urllib.urlencode(param)

    request  = urllib2.Request(url,data,header_params)
    gcontext = ssl.SSLContext(ssl.PROTOCOL_TLSv1)
    response = urllib2.urlopen(request,context=gcontext)


    ### write to file 
    filename = "/home/dan/tabletest.csv"
    f = open(filename,'w+')
    f.write(response.read())
    # f.write("testing")
    f.close()


